src/services/urlUtils.ts

```diff
@@ -122,41 +122,59 @@ namespace UrlUtils {
     });
   };
 
-  export const getNavigations = (request: UrlBeautifier.SearchUrlState) => {
+  // tslint:disable-next-line max-line-length
+  export const getNavigations = (state: Store.Indexed<Store.Navigation> | Store.AvailableNavigations, request: UrlBeautifier.SearchUrlState) => {
     return {
-      allIds: getAllIds(request),
-      byId: getById(request),
+      allIds: getAllIds(state, request),
+      byId: getById(state, request),
     };
   };
 
-  export const getAllIds = (request: UrlBeautifier.SearchUrlState) => {
-    return request.refinements.reduce((fields, refinement) => {
-      if (!fields.includes(refinement.field)) {
-        fields.push(refinement.field);
+  // tslint:disable-next-line max-line-length
+  export const getAllIds = (state: Store.Indexed<Store.Navigation> | Store.AvailableNavigations, request: UrlBeautifier.SearchUrlState) => {
+    const allIds = state.allIds;
+    request.refinements.forEach((refinement) => {
+      if (!allIds.includes(refinement.field)) {
+        allIds.push(refinement.field);
       }
-      return fields;
-    }, []);
+    });
+
+    return allIds;
   };
 
-  export const getById = (request: UrlBeautifier.SearchUrlState) => {
-    return request.refinements.reduce((navigations, refinement) => {
+  // tslint:disable-next-line max-line-length
+  export const getById = (state: Store.Indexed<Store.Navigation> | Store.AvailableNavigations, request: UrlBeautifier.SearchUrlState) => {
+    const byId = state.byId;
+
+    request.refinements.forEach((refinement) => {
       const field = refinement.field;
       const transformed =
         'low' in refinement ? { low: refinement['low'], high: refinement['high'] } : { value: refinement['value'] };
-      if (field in navigations) {
-        const navigation = navigations[field];
-        navigation.selected.push(navigation.refinements.push(transformed) - 1);
+      if (byId[field]) {
+        const navigation = byId[field];
+        const existingIndex = navigation.refinements
+          .findIndex((ref) =>
+            Adapters.Search.refinementsMatch(<any>transformed, <any>ref, navigation.range ? 'Range' : 'Value')
+          );
+        if (existingIndex !== -1) {
```

Comment:
Consider inverting the condition and check `existingIndex === 1` first, then do an `else if` clause for the `includes` case to remove a level of nesting.